spec/helpers/http_mock.lua

```diff
@@ -0,0 +1,176 @@
+    access = [[
+      ngx.req.set_header("X-Test", "test")
+      ngx.print("ok")
+      ngx.exit(200)
+    ]],
+  },
+}
+
+local function default_field(tbl, key, default)
+  if tbl[key] == nil then
+    tbl[key] = default
+  end
+end
+
+-- create a mock instance which represents a HTTP mocking server
+-- @param listens: the listen directive of the mock server, default to "0.0.0.0:8000"
+-- @param code: the code of the mock server, default to a simple response.
+-- @param opts: options for the mock server, left it empty to use the defaults
+-- @return: a mock instance
+-- @usage
+-- local mock = http_mock.new(8000, [[
+--   ngx.req.set_header("X-Test", "test")
+--   ngx.print("hello world")
+-- ]],  {
+--   prefix = "mockserver",
+--   record_opts = {
+--     resp = true,
+--     resp_body = true,
+--   },
+--   tls = true,
+-- })
+--
+-- mock:start()
+-- local client = mock:get_client() -- get a client to access the mocking port
+-- local res = assert(client:send({}))
+-- assert.response(res).has.status(200)
+-- assert.response(res).has.header("X-Test", "test")
+-- assert.response(res).has.body("hello world")
+-- mock.eventually:has_response(function(resp)
+--   assert.same(resp.body, "hello world")
+-- end)
+-- mock:wait_until_no_request() -- wait until all the requests are finished
+-- mock:clean() -- clean the records
+-- client:send({})
+-- client:send({})
+-- local records = mock:get_record() -- get all the records of HTTP sessions
+-- mock:stop()
+-- 
+-- listens can be a number, which will be used as the port of the mock server;
+-- or a string, which will be used as the param of listen directive of the mock server;
+-- or a table represents multiple listen ports.
+-- if the port is not specified, a random port will be used.
+-- if the port is a number and opts.tls is set to ture, ssl will be appended.
+--
+-- routes can be a table like this:
+-- routes = {
+--   ["/"] = {
+--     access = [[
+--       ngx.req.set_header("X-Test", "test")
+--       ngx.print("hello world")
+--     ]],
+--     log = [[
+--       ngx.log(ngx.ERR, "log test!")
+--     ]],
+--     directives = {
+--       "rewrite ^/foo /bar break;",
+--     },
+--   },
+-- }
+-- or a string, which will be used as the access phase handler.
+--
+-- opts:
+-- prefix: the prefix of the mock server, default to "mockserver"
+-- hostname: the hostname of the mock server, default to "_"
+-- directives: the extra directives of the mock server, default to {}
+-- record_opts: the options for recording with fields listed below:
+--   collect_req: whether to record request, default to true
+--   collect_req_body_large: whether to record large request body by setting, default to true
+--   collect_resp: whether to record response, default to false
+--   collect_resp_body: whether to record response body by setting, default to false
+--   collect_err: whether to record error, default to true
+-- tls: whether to use tls, default to false
+function http_mock.new(listens, code, opts)
```

Comment:
`code` could better be named `routes` as that is how the field is passed to nginx anyway.